parsons/databases/bigquery/bigquery.py

```diff
@@ -0,0 +1,1123 @@
+        """
+
+        with self.connection() as connection:
+            return self.query_with_connection(
+                sql, connection, parameters=parameters, return_values=return_values
+            )
+
+    def query_with_connection(
+        self, sql, connection, parameters=None, commit=True, return_values: bool = True
+    ):
+        """
+        Execute a query against the BigQuery database, with an existing connection.
+        Useful for batching queries together. Will return ``None`` if the query
+        returns zero rows.
+
+        `Args:`
+            sql: str
+                A valid SQL statement
+            connection: obj
+                A connection object obtained from ``redshift.connection()``
+            parameters: list
+                A list of python variables to be converted into SQL values in your query
+            commit: boolean
+                Must be true. BigQuery
+
+        `Returns:`
+            Parsons Table
+                See :ref:`parsons-table` for output options.
+        """
+
+        if not commit:
+            raise ValueError(
+                """
+                BigQuery implementation uses an API client which always auto-commits. If you wish to wrap
+                multiple queries in a transaction, use Mulit-Statement transactions within a single query
+                as outlined here: https://cloud.google.com/bigquery/docs/transactions or use the
+                `query_with_transaction` method on the BigQuery connector.
+            """
+            )
+
+        # get our connection and cursor
+        with self.cursor(connection) as cursor:
+            # Run the query
+            cursor.execute(sql, parameters)
+
+            if not return_values:
+                return None
+
+            final_table = self._fetch_query_results(cursor=cursor)
+
+            return final_table
+
+    def query_with_transaction(self, queries, parameters=None):
+        queries_with_semicolons = [ends_with_semicolon(q) for q in queries]
+        queries_on_newlines = "\n".join(queries_with_semicolons)
+        queries_wrapped = f"""
+        BEGIN
+            BEGIN TRANSACTION;
+            
+            {queries_on_newlines}
+
+            COMMIT TRANSACTION;
+
+            EXCEPTION WHEN ERROR THEN
+            -- Roll back the transaction inside the exception handler.
+            SELECT @@error.message;
+            ROLLBACK TRANSACTION;
+        END;
+        """
+        self.query(sql=queries_wrapped, parameters=parameters, return_values=False)
+
+    def copy_from_gcs(
+        self,
+        gcs_blob_uri: str,
+        table_name: str,
+        if_exists: str = "fail",
+        max_errors: int = 0,
+        data_type: str = "csv",
+        csv_delimiter: str = ",",
+        ignoreheader: int = 1,
+        nullas: str = None,
+        allow_quoted_newlines: bool = True,
+        allow_jagged_rows: bool = True,
+        quote: str = None,
+        schema: list = None,
```

Comment:
This type annotation should be `Optional[list]`, or better yet `Optional[List[dict]]`